Disable host name collision prevention

bfallonf opened the pull request:
Card: https://trello.com/c/GP4rxvxy/431-document-prevent-changes-to-hosts-defined-in-ingress-resources-ingress

Information over email. @JacobTanenbaum , can I ask for a tech preview? I'm wondering if I've mistaken any of the language. Thanks!

JacobTanenbaum commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,34 @@ can now access any pods and services in the cluster and vice versa.
 `--selector=<project_selector>` option.
 
+[[admin-guide-disabling-hostname-collision]]
+== Disabling Host Name Collision Detection For Routes
```
this is for ingress objects not for routes

JacobTanenbaum commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,34 @@ can now access any pods and services in the cluster and vice versa.
+[[admin-guide-disabling-hostname-collision]]
+== Disabling Host Name Collision Detection For Routes
+
+In {product-title}, host name collision prevention is enabled by default. This means that if a route is set to a host name, it will not be changed. When disabled, {product-title} uses the route creation timestamp to determine the oldest route for a host name, then uses that route. If an older route changes its host name, then it can hijack a host name with a newer route. 
```
nit: it will not be changed. -> it can not be changed.

bfallonf commented:
Thanks @JacobTanenbaum . I've changed as suggested. Can I ask what other types of ingress objects there are then? I changed it a little so that routes are an example of this, but it would probably be better to list them? Unless I'm misunderstanding the whole concept here. Thanks.

knobunc commented:
@bfallonf no worries.  Route Objects vs. Ingress Objects are confusing.

The Route Objects are the original OpenShift implementation of the concept.  Kubernetes initially didn't have the concept and didn't see the need.  But when Routes were popular, and useful, they decided to implement the concept... but in a slightly different way.  So Kubernetes added "Ingress Objects".  We added support to our router to support Ingress Objects as well as Route Objects in 1.5 (but ingress support is disabled by default).

You can 'oc get ingress' and 'oc get routes'.

knobunc commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,40 @@ can now access any pods and services in the cluster and vice versa.
+== Disabling Host Name Collision Detection For Ingress Objects
+
+In {product-title}, host name collision prevention is enabled by default. This
+means that if an ingress object, for example, a route, is set to a host name, it
```
With routes we forbid the host update.  There is no way to disable it.  This just adds the security to ingress that routes already have (it is necessary for a multi-tenant, self-service, environment... but other users don't need this)

knobunc commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,40 @@ can now access any pods and services in the cluster and vice versa.
+If an older route changes its host name, then it can hijack a host name with a
+newer route. 
+
+However, disabling host name collision prevention can enable upstream Kubernetes
```
s/can enable/enables/   ?

bfallonf commented:
@JacobTanenbaum @knobunc Thanks for the suggestions. I've given this another go. Can I ask for another review? I'm hoping that I've properly shown the reason for this section. Thanks!

knobunc commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,38 @@ can now access any pods and services in the cluster and vice versa.
+route or ingress object. When disabled, {product-title} uses the route creation
+timestamp to determine the oldest route for a host name, then uses that route.
+If an older route changes its host name, or if an ingress object is introduced,
+then it can hijack a host name with a newer route. 
```
How about:

In {product-title}, host name collision prevention is enabled by default for both routes and ingress objects. This means that the host name in a route or ingress object can only be set on creation and not edited afterwards. When disabled, {product-title} will allow you to edit the host name after creation  Because {product-title} uses the route creation timestamp to determine the oldest route or ingress object for a given host name, then uses that object if an older route changes its host name, or if an ingress object is introduced, then it can hijack a host name with a newer route.

Most installations will not need to disable this, but in Kubernetes the host names in ingress objects are allowed to be edited, and if you depend upon that behavior then you can disable the safety checks.

bfallonf commented:
Thanks @knobunc . I've changed to what you suggested, with some rewording.

JacobTanenbaum commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,42 @@ can now access any pods and services in the cluster and vice versa.
+objects is enabled by default. This means that the host name in a route or
+ingress object can only be set on creation and not edited afterwards. Disabling
+host name collision prevention lets you edit a host name after creation.
+However, because {product-title} uses the route creation timestamp to determine
```
uses the object creation timestamp

JacobTanenbaum commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,42 @@ can now access any pods and services in the cluster and vice versa.
+This is relevant to {product-title} installations that depend upon Kubernetes
+behavior, including allowing the host names in ingress objects be edited. You
+can disable host name collision to enable system-wide routing through ingress
+objects.
```
this does not enable routing through ingress objects just editing of the hostname field

bfallonf replied:
Sure. I deleted the sentence.

JacobTanenbaum commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,42 @@ can now access any pods and services in the cluster and vice versa.
+In {product-title}, host name collision prevention for routes and ingress
+objects is enabled by default. This means that the host name in a route or
+ingress object can only be set on creation and not edited afterwards. Disabling
+host name collision prevention lets you edit a host name after creation.
```
It might be important to mention that this only enables editing of hostnames for ingress objects. This controller has no effect on route objects

bfallonf commented:
Thanks @JacobTanenbaum . I've made the changes that you suggested. I'll move this along.

@gaurav-nelson Peer review please?

gaurav-nelson commented on admin_guide/managing_networking.adoc:
```diff
@@ -85,6 +85,40 @@ can now access any pods and services in the cluster and vice versa.
+
+. Add the following to the `master.yaml` file:
++
+----
```
Are we moving away from putting in [source, yaml] tags?

bfallonf replied:
Oh. Sure. Can put in.

gaurav-nelson commented:
@bfallonf Everything looks good! :100:

bfallonf commented:
[rev_history]
|xref:../admin_guide/managing_networking.adoc#admin-guide-manage-networking[Managing Networking]
|Added the xref:../admin_guide/managing_networking.adoc#admin-guide-disabling-hostname-collision[Disabling Host Name Collision Prevention For Ingress Objects] section.
%

bfallonf commented:
Thanks @gaurav-nelson . I'll merge.
